lib/spack/docs/tutorial_advanced_packaging.rst

```diff
@@ -0,0 +1,560 @@
+  arpack-ng@3.5.0  hdf5@1.10.1   libpciaccess@0.13.5  libtool@2.4.6  m4@1.4.18  ncurses@6.0          openblas@0.2.20  openssl@1.0.2k     superlu@5.2.1       xz@5.2.3
+  cmake@3.9.4      hwloc@1.11.8  libsigsegv@2.11      libxml2@2.9.4  mpich@3.2  netlib-lapack@3.6.1  openmpi@3.0.0    pkg-config@0.29.2  util-macros@1.19.1  zlib@1.2.11
+
+to help focusing on the topic at hand and not on configuration issues.
+If you already started the image, then you can set your preferred editor
+(possible choices are ``emacs``, ``vim``, ``nano``):
+
+.. code-block:: console
+
+  root@advanced-packaging-tutorial:/# export EDITOR=emacs # or vim, or nano
+
+and move directly to :ref:`specs_build_interface_tutorial`.
+
+If you need instead to work in a local folder, the steps to set it up are:
+
+.. code-block:: console
+
+  $ git clone https://github.com/spack/spack.git
+  Cloning into 'spack'...
+  remote: Counting objects: 92731, done.
+  remote: Compressing objects: 100% (1108/1108), done.
+  remote: Total 92731 (delta 1964), reused 4186 (delta 1637), pack-reused 87932
+  Receiving objects: 100% (92731/92731), 33.31 MiB | 64.00 KiB/s, done.
+  Resolving deltas: 100% (43557/43557), done.
+  Checking connectivity... done.
+
+  $ cd spack
+  $ git checkout tutorials/advanced_packaging
+  Branch tutorials/advanced_packaging set up to track remote branch tutorials/advanced_packaging from origin.
+  Switched to a new branch 'tutorials/advanced_packaging'
+
+At this point you can install the software that will be used
+during the rest of the tutorial (the output of the commands is omitted
+for the sake of brevity):
+
+.. code-block:: console
+
+  $ spack install openblas
+  $ spack install netlib-lapack
+  $ spack install mpich
+  $ spack install openmpi
+  $ spack install --only=dependencies armadillo ^openblas
+  $ spack install --only=dependencies netcdf
+  $ spack install --only=dependencies elpa
+
+Now, you are ready to set your preferred ``EDITOR`` and continue with
+the rest of the tutorial.
+
+
+.. _specs_build_interface_tutorial:
+
+----------------------
+Spec's build interface
+----------------------
+
+Spack is designed with a particular care on assigning responsibilities
+to the right entities, as this results in a clearer and more intuitive interface
+for the users.
+When it comes to packaging, one of the most fundamental guideline that
+emerged from this tenet is that:
+
+  *It's a package responsibility to know
+  every software it directly depends on and to expose to others how to
+  use the services it provides*.
+
+Spec's build interface is a protocol-like implementation of this guideline
+that allows packages to easily query their dependencies,
+and prescribes how they should expose their own build information.
+
+^^^^^^^^^^^^^^^^^^^^
+A motivating example
+^^^^^^^^^^^^^^^^^^^^
+
+Suppose we need to package ``armadillo``. This library has a few dependencies,
+including LAPACK. Let's open an editor and complete the stub that
```

Comment:
"Let's open..." ->

"We've started work on a package for armadillo, you should open it, read through the comment that starts with "TUTORIAL:" and complete the `cmake_args` section (replacing the FIXME's with appropriate bits from the `specs` list.